engine/src/main/java/org/hibernate/search/engine/backend/work/execution/OperationSubmitter.java

```diff
@@ -118,20 +118,53 @@ private OffloadingExecutorOperationSubmitter(Consumer<Runnable> executor) {
 		}
 
 		@Override
-		public <T> void submitToQueue(BlockingQueue<? super T> queue, T element, Consumer<? super T> blockingRetryProducer) {
+		public <T> void submitToQueue(BlockingQueue<? super T> queue, T element, Consumer<? super T> blockingRetryProducer,
+				Consumer<Throwable> asyncFailureReporter) {
 			if ( !queue.offer( element ) ) {
-				executor.accept( () -> blockingRetryProducer.accept( element ) );
+				try {
+					this.executor.accept( new RetryAction<>( element, blockingRetryProducer, asyncFailureReporter ) );
+				}
+				catch (Throwable throwable) {
+					asyncFailureReporter.accept( throwable );
+				}
 			}
 		}
 
 		@Override
 		public <T extends Runnable> void submitToExecutor(SimpleScheduledExecutor executor, T element,
-				Consumer<? super T> blockingRetryProducer) {
+				Consumer<? super T> blockingRetryProducer, Consumer<Throwable> asyncFailureReporter) {
 			try {
 				executor.offer( element );
 			}
 			catch (RejectedExecutionException e) {
-				this.executor.accept( () -> blockingRetryProducer.accept( element ) );
+				try {
+					this.executor.accept( new RetryAction<>( element, blockingRetryProducer, asyncFailureReporter ) );
+				}
+				catch (Throwable throwable) {
+					asyncFailureReporter.accept( throwable );
+				}
```

Comment:
I think you can leave this out. Callers will want to catch synchronous failures as usual.